librad/src/meta/project.rs

```diff
@@ -184,4 +414,231 @@ pub mod tests {
+        assert_eq!(1, prj.signatures.iter().filter(|s| s.key == m1).count());
+        assert_eq!(1, prj.signatures.iter().filter(|s| s.key == m2).count());
+
+        // Check signature verification
+        let data = prj.canonical_data().unwrap();
+        for s in prj.signatures.iter() {
+            let sig = Signature::from_hex_string(&s.sig).unwrap();
+            assert_eq!(sig.verify(&data, s.key.device_key()), true);
+        }
+
+        // Check signature removal
+        prj.remove_signature(&m1);
+        assert_eq!(2, prj.signatures.len());
+    }
+
+    #[test]
+    fn test_project_verification() {
+        let (m0, k0) = new_peer_with_key(42);
+        let (m1, k1) = new_peer_with_key(1);
+        let (m2, k2) = new_peer_with_key(2);
+
+        // A new project is not valid because the owner has not signed it
+        let mut prj = Project::new("foo", &m0);
+        assert!(matches!(prj.check_validity(), Err(Error::SignatureMissing)));
+        assert!(!prj.is_valid());
+        // Adding the signature fixes it
+        prj.add_signature(&k0).unwrap();
+        assert!(matches!(prj.check_validity(), Ok(())));
+        assert!(prj.is_valid());
+        // Adding maintainers without signatures invalidates it
+        prj.add_maintainer(&m1);
+        prj.add_maintainer(&m2);
+        assert!(matches!(prj.check_validity(), Err(Error::SignatureMissing)));
+        // Adding the missing signatures does not fix it: m0 signed a previous revision
+        prj.add_signature(&k1).unwrap();
+        prj.add_signature(&k2).unwrap();
+        assert!(matches!(
+            prj.check_validity(),
+            Err(Error::SignatureVerificationFailed)
+        ));
+        // Cannot sign a project twice with the sme key
+        assert!(matches!(
+            prj.add_signature(&k0),
+            Err(Error::SignatureAlreadyPresent)
+        ));
+        // Removing the signature and re adding it fixes the project
+        prj.remove_signature(&m0);
+        prj.add_signature(&k0).unwrap();
+        assert!(prj.is_valid());
+        // Removing a maintainer invalidates it again
+        prj.maintainers.pop();
+        assert!(matches!(
+            prj.check_validity(),
+            Err(Error::SignatureFromNonMaintainer)
+        ));
+    }
+    #[test]
+    fn test_project_update() {
+        let (m0, k0) = new_peer_with_key(42);
+        let (m1, k1) = new_peer_with_key(1);
+        let (m2, k2) = new_peer_with_key(2);
+        let (m3, k3) = new_peer_with_key(3);
+        let (m4, k4) = new_peer_with_key(4);
+
+        // Empty history is invalid
+        let mut history: Vec<Project> = vec![];
+        assert!(matches!(
+            Project::check_history(history.iter()),
+            Err(HistoryVerificationError::EmptyHistory)
+        ));
+        // History with invalid project is invalid
+        let mut prj = Project::new("foo", &m0);
+        prj.revision = 1;
+        history.push(prj);
+        assert!(matches!(
```

Comment:
Same here, and a few other places

Reply:
I guess this would require a `#[derive(Eq)]` on all errors.
I am not opposed to that in principle but usually this is not needed and errors are checked with `match` clauses and not comparisons...

Reply:
We lose nothing by deriving `Eq` but we gain diffing at test time :) `assert` will just say that the bool was false, whereas when/if the test fails with `assert_eq` then we see what the result of `check_history` was and can inspect what went wrong.

Reply:
The problem is that _deriving_ `Eq` is impossible: `serde_json::error::Error` and `std::string::FromUtf8Error`, which we don't control, do not implement `Eq`.

_Implementing_ `Eq` could be doable, but with all the problems involved in deep equality when leafs cannot be comparable... IMHO it's not worth it.

When there's a problem in a test I just put a `println` of the error just before the failing line, rerun the test, and see what happened. A bit less than ideal but practical enough.

Reply:
Ach... Fair enough :)